File: src/annoy.cpp
// -*- mode: C++; c-indent-level: 4; c-basic-offset: 4; indent-tabs-mode: nil; -*-
//
//  RcppAnnoy -- Rcpp bindings to Annoy library for Approximate Nearest Neighbours
//
//  Copyright (C) 2014 - 2020  Dirk Eddelbuettel
//
//  This file is part of RcppAnnoy
//
//  RcppAnnoy is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, either version 2 of the License, or
//  (at your option) any later version.
//
//  RcppAnnoy is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with RcppAnnoy.  If not, see <http://www.gnu.org/licenses/>.

// simple C++ modules to wrap to two templated classes from Annoy
//
// uses annoylib.h (from Annoy) and provides R access via Rcpp
//
// Dirk Eddelbuettel, Nov 2014

#include "RcppAnnoy.h"

template< typename S, typename T, typename Distance, typename Random, class ThreadedBuildPolicy >
class Annoy
{
protected:
    AnnoyIndex<S, T, Distance, Random, ThreadedBuildPolicy> *ptr;
    unsigned int vectorsz;
public:
    Annoy(int n) : vectorsz(n) {
        ptr = new AnnoyIndex<S, T, Distance, Random, ThreadedBuildPolicy>(n);
    }
    ~Annoy() { if (ptr != NULL) delete ptr; }
    void addItem(S item, Rcpp::NumericVector dv) {
        if (item < 0) Rcpp::stop("Inadmissible item value %d", item);
        std::vector<T> fv(dv.size());
        std::copy(dv.begin(), dv.end(), fv.begin());
        char *errormsg;
        if (!ptr->add_item(item, &fv[0], &errormsg)) Rcpp::stop(errormsg);
    }
    void   callBuild(int n)               { ptr->build(n);                  }
    void   callUnbuild()                  { ptr->unbuild();                 }
    void   callSave(std::string filename) { ptr->save(filename.c_str());    }
    void   callLoad(std::string filename) { ptr->load(filename.c_str());    }
    void   callUnload()                   { ptr->unload();                  }
    int    getNItems()                    { return ptr->get_n_items();      }
    int    getNTrees()                    { return ptr->get_n_trees();      }
    double getDistance(int i, int j)      { return ptr->get_distance(i, j); }
    void   verbose(bool v)                { ptr->verbose(v);                }
    void   setSeed(int s)                 { ptr->set_seed(s);               }

    std::vector<S> getNNsByItem(S item, size_t n) {
        std::vector<S> result;
        ptr->get_nns_by_item(item, n, -1, &result, NULL);
        return result;
    }

    Rcpp::List getNNsByItemList(S item, size_t n, int search_k, bool include_distances) {
        if (include_distances) {
            std::vector<S> result;
            std::vector<T> distances;
            ptr->get_nns_by_item(item, n, search_k, &result, &distances);
            return Rcpp::List::create(Rcpp::Named("item")     = result,
                                      Rcpp::Named("distance") = distances);
        } else {
            std::vector<S> result;
            ptr->get_nns_by_item(item, n, search_k, &result, NULL);
            return Rcpp::List::create(Rcpp::Named("item") = result);
        }
    }

    std::vector<S> getNNsByVector(std::vector<double> dv, size_t n) {
        std::vector<T> fv(dv.size());
        std::copy(dv.begin(), dv.end(), fv.begin());
        std::vector<S> result;
        ptr->get_nns_by_vector(&fv[0], n, -1, &result, NULL);
        return result;
    }

    Rcpp::List getNNsByVectorList(std::vector<T> fv, size_t n, int search_k, bool include_distances) {
        if (fv.size() != vectorsz) {
            Rcpp::stop("fv.size() != vector_size");
        }
        if (include_distances) {
            std::vector<S> result;
            std::vector<T> distances;
            ptr->get_nns_by_vector(&fv[0], n, search_k, &result, &distances);
            return Rcpp::List::create(
                Rcpp::Named("item") = result,
                Rcpp::Named("distance") = distances);
        } else {
            std::vector<S> result;
            ptr->get_nns_by_vector(&fv[0], n, search_k, &result, NULL);
            return Rcpp::List::create(Rcpp::Named("item") = result);
        }
    }

    std::vector<double> getItemsVector(S item) {
        std::vector<T> fv(vectorsz);
        ptr->get_item(item, &fv[0]);
        std::vector<double> dv(fv.size());
        std::copy(fv.begin(), fv.end(), dv.begin());
        return dv;
    }

    bool onDiskBuild(std::string fname) {
        char *errormsg;
        if (!ptr->on_disk_build(fname.c_str(), &errormsg)) Rcpp::stop(errormsg);
        return true;
    }

};

typedef Annoy<int32_t, float,    Angular,   Kiss64Random, RcppAnnoyIndexThreadPolicy> AnnoyAngular;
typedef Annoy<int32_t, float,    Euclidean, Kiss64Random, RcppAnnoyIndexThreadPolicy> AnnoyEuclidean;
typedef Annoy<int32_t, float,    Manhattan, Kiss64Random, RcppAnnoyIndexThreadPolicy> AnnoyManhattan;
typedef Annoy<int32_t, uint64_t, Hamming,   Kiss64Random, RcppAnnoyIndexThreadPolicy> AnnoyHamming;

RCPP_EXPOSED_CLASS_NODECL(AnnoyAngular)
RCPP_MODULE(AnnoyAngular) {
    Rcpp::class_<AnnoyAngular>("AnnoyAngular")

        .constructor<int32_t>("constructor with integer count")

        .method("addItem",        &AnnoyAngular::addItem,         "add item")
        .method("build",          &AnnoyAngular::callBuild,       "build an index")
        .method("unbuild",        &AnnoyAngular::callUnbuild,     "unbuild an index")
        .method("save",           &AnnoyAngular::callSave,        "save index to file")
        .method("load",           &AnnoyAngular::callLoad,        "load index from file")
        .method("unload",         &AnnoyAngular::callUnload,      "unload index")
        .method("getDistance",    &AnnoyAngular::getDistance,     "get distance between i and j")
        .method("getNNsByItem",   &AnnoyAngular::getNNsByItem,
                "retrieve Nearest Neigbours given item")
        .method("getNNsByItemList",  &AnnoyAngular::getNNsByItemList,
                "retrieve Nearest Neigbours given item")
        .method("getNNsByVector", &AnnoyAngular::getNNsByVector,
                "retrieve Nearest Neigbours given vector")
        .method("getNNsByVectorList",  &AnnoyAngular::getNNsByVectorList,
                "retrieve Nearest Neigbours given vector")
        .method("getItemsVector", &AnnoyAngular::getItemsVector,  "retrieve item vector")
        .method("getNItems",      &AnnoyAngular::getNItems,       "get number of items")
        .method("getNTrees",      &AnnoyAngular::getNTrees,       "get number of trees")
        .method("setVerbose",     &AnnoyAngular::verbose,         "set verbose")
        .method("setSeed",        &AnnoyAngular::setSeed,         "set seed")
        .method("onDiskBuild",    &AnnoyAngular::onDiskBuild,     "build in given file")
        ;
}

RCPP_EXPOSED_CLASS_NODECL(AnnoyEuclidean)
RCPP_MODULE(AnnoyEuclidean) {
    Rcpp::class_<AnnoyEuclidean>("AnnoyEuclidean")

        .constructor<int32_t>("constructor with integer count")

        .method("addItem",        &AnnoyEuclidean::addItem,        "add item")
        .method("build",          &AnnoyEuclidean::callBuild,      "build an index")
        .method("unbuild",        &AnnoyEuclidean::callUnbuild,    "unbuild an index")
        .method("save",           &AnnoyEuclidean::callSave,       "save index to file")
        .method("load",           &AnnoyEuclidean::callLoad,       "load index from file")
        .method("unload",         &AnnoyEuclidean::callUnload,     "unload index")
        .method("getDistance",    &AnnoyEuclidean::getDistance,    "get distance between i and j")
        .method("getNNsByItem",   &AnnoyEuclidean::getNNsByItem,
                "retrieve Nearest Neigbours given item")
        .method("getNNsByItemList",  &AnnoyEuclidean::getNNsByItemList,
                "retrieve Nearest Neigbours given item")
        .method("getNNsByVector", &AnnoyEuclidean::getNNsByVector,
                "retrieve Nearest Neigbours given vector")
        .method("getNNsByVectorList",&AnnoyEuclidean::getNNsByVectorList,
                "retrieve Nearest Neigbours given vector")
        .method("getItemsVector", &AnnoyEuclidean::getItemsVector, "retrieve item vector")
        .method("getNItems",      &AnnoyEuclidean::getNItems,      "get number of items")
        .method("getNTrees",      &AnnoyEuclidean::getNTrees,      "get number of trees")
        .method("setVerbose",     &AnnoyEuclidean::verbose,        "set verbose")
        .method("setSeed",        &AnnoyEuclidean::setSeed,        "set seed")
        .method("onDiskBuild",    &AnnoyEuclidean::onDiskBuild,    "build in given file")
        ;
}

RCPP_EXPOSED_CLASS_NODECL(AnnoyManhattan)
RCPP_MODULE(AnnoyManhattan) {
    Rcpp::class_<AnnoyManhattan>("AnnoyManhattan")

        .constructor<int32_t>("constructor with integer count")

        .method("addItem",        &AnnoyManhattan::addItem,        "add item")
        .method("build",          &AnnoyManhattan::callBuild,      "build an index")
        .method("unbuild",        &AnnoyManhattan::callUnbuild,    "unbuild an index")
        .method("save",           &AnnoyManhattan::callSave,       "save index to file")
        .method("load",           &AnnoyManhattan::callLoad,       "load index from file")
        .method("unload",         &AnnoyManhattan::callUnload,     "unload index")
        .method("getDistance",    &AnnoyManhattan::getDistance,    "get distance between i and j")
        .method("getNNsByItem",   &AnnoyManhattan::getNNsByItem,
                "retrieve Nearest Neigbours given item")
        .method("getNNsByItemList",  &AnnoyManhattan::getNNsByItemList,
                "retrieve Nearest Neigbours given item")
        .method("getNNsByVector", &AnnoyManhattan::getNNsByVector,
                "retrieve Nearest Neigbours given vector")
        .method("getNNsByVectorList",&AnnoyManhattan::getNNsByVectorList,
                "retrieve Nearest Neigbours given vector")
        .method("getItemsVector", &AnnoyManhattan::getItemsVector, "retrieve item vector")
        .method("getNItems",      &AnnoyManhattan::getNItems,      "get number of items")
        .method("getNTrees",      &AnnoyManhattan::getNTrees,      "get number of trees")
        .method("setVerbose",     &AnnoyManhattan::verbose,        "set verbose")
        .method("setSeed",        &AnnoyManhattan::setSeed,        "set seed")
        .method("onDiskBuild",    &AnnoyManhattan::onDiskBuild,    "build in given file")
        ;
}

RCPP_EXPOSED_CLASS_NODECL(AnnoyHamming)
RCPP_MODULE(AnnoyHamming) {
    Rcpp::class_<AnnoyHamming>("AnnoyHamming")

        .constructor<int32_t>("constructor with integer count")

        .method("addItem",        &AnnoyHamming::addItem,        "add item")
        .method("build",          &AnnoyHamming::callBuild,      "build an index")
        .method("unbuild",        &AnnoyHamming::callUnbuild,    "unbuild an index")
        .method("save",           &AnnoyHamming::callSave,       "save index to file")
        .method("load",           &AnnoyHamming::callLoad,       "load index from file")
        .method("unload",         &AnnoyHamming::callUnload,     "unload index")
        .method("getDistance",    &AnnoyHamming::getDistance,    "get distance between i and j")
        .method("getNNsByItem",   &AnnoyHamming::getNNsByItem,
                "retrieve Nearest Neigbours given item")
        .method("getNNsByItemList",  &AnnoyHamming::getNNsByItemList,
                "retrieve Nearest Neigbours given item")
        .method("getNNsByVector", &AnnoyHamming::getNNsByVector,
                "retrieve Nearest Neigbours given vector")
        .method("getNNsByVectorList",&AnnoyHamming::getNNsByVectorList,
                "retrieve Nearest Neigbours given vector")
        .method("getItemsVector", &AnnoyHamming::getItemsVector, "retrieve item vector")
        .method("getNItems",      &AnnoyHamming::getNItems,      "get number of items")
        .method("getNTrees",      &AnnoyHamming::getNTrees,      "get number of trees")
        .method("setVerbose",     &AnnoyHamming::verbose,        "set verbose")
        .method("setSeed",        &AnnoyHamming::setSeed,        "set seed")
        .method("onDiskBuild",    &AnnoyHamming::onDiskBuild,    "build in given file")
        ;
}

File: R/annoy.R
#' @name AnnoyIndex
#'
#' @aliases
#' AnnoyEuclidean Rcpp_AnnoyEuclidean-class Rcpp_AnnoyEuclidean
#' AnnoyAngular   Rcpp_AnnoyAngular-class   Rcpp_AnnoyAngular
#' AnnoyManhattan Rcpp_AnnoyManhattan-class Rcpp_AnnoyManhattan
#' AnnoyHamming   Rcpp_AnnoyHamming-class   Rcpp_AnnoyHamming
#'
#' @title Approximate Nearest Neighbors with Annoy
#'
#' @description
#' Annoy is a small library written to provide fast and memory-efficient
#' nearest neighbor lookup from a possibly static index which can be
#' shared across processes.
#'
#' @section Usage:
#' \preformatted{
#' a <- new(AnnoyEuclidean, vectorsz)
#'
#' a$setSeed(0)
#' a$setVerbose(0)
#'
#' a$addItem(i, dv)
#'
#' a$getNItems()
#'
#' a$getItemsVector(i)
#' a$getDistance(i, j)
#'
#' a$build(n_trees)
#'
#' a$getNNsByItem(i, n)
#' a$getNNsByItemList(i, n, search_k, include_distances)
#'
#' a$getNNsByVector(v, n)
#' a$getNNsByVectorList(v, n, search_k, include_distances)
#'
#' a$save(fn)
#' a$load(fn)
#' a$unload()
#' }
#'
#' @section Details:
#'
#' \code{new(Class, vectorsz)}
#' Create a new Annoy instance of type \code{Class} where \code{Class}
#' is on of the following:
#' \code{AnnoyEuclidean},
#' \code{AnnoyAngular},
#' \code{AnnoyManhattan},
#' \code{AnnoyHamming}.
#' \code{vectorsz} denotes the length of the vectors that the Annoy instance
#' will be indexing.
#'
#' \code{$addItem(i, v)}
#' Adds item \code{i} (any nonnegative integer) with vector \code{v}.
#' Note that it will allocate memory for \code{max(i) + 1} items.
#'
#' \code{$build(n_trees)}
#' Builds a forest of \code{n_trees} trees.
#' More trees gives higher precision when querying.
#' After calling \code{build}, no more items can be added.
#'
#' \code{$save(fn)}
#' Saves the index to disk as filename \code{fn}.
#' After saving, no more items can be added.
#'
#' \code{$load(fn)}
#' Loads (mmaps) an index from filename \code{fn} on disk.
#'
#' \code{$unload()}
#' Unloads index.
#'
#' \code{$getDistance(i, j)}
#' Returns the distance between items \code{i} and \code{j}
#'
#' \code{$getNNsByItem(i, n)}
#' Returns the \code{n} closest items as an integer vector of indices.
#'
#' \code{$getNNsByVector(v, n)}
#' Same as \code{$getNNsByItem}, but queries by vector \code{v} rather than
#' index \code{i}.
#'
#' \code{$getNNsByItemList(i, n, search_k = -1, include_distances = FALSE)}
#' Returns the n closest items to item \code{i} as a list.
#' During the query it will inspect up to \code{search_k} nodes which
#' defaults to \code{n_trees * n} if not provided.
#' \code{search_k} gives you a run-time tradeoff between better accuracy and
#' speed.
#' If you set \code{include_distances} to \code{TRUE},
#' it will return a length 2 list with elements \code{"item"} &
#' \code{"distance"}.
#' The \code{"item"} element contains the \code{n} closest items as an integer
#' vector of indices.
#' The optional \code{"distance"} element contains the corresponding distances
#' to \code{"item"} as a numeric vector.
#'
#' \code{$getNNsByVectorList(i, n, search_k = -1, include_distances = FALSE)}
#' Same as \code{$getNNsByItemList}, but queries by vector \code{v} rather than
#' index \code{i}
#'
#' \code{$getItemsVector(i)}
#' Returns the vector for item \code{i} that was previously added.
#'
#' \code{$getNItems()}
#' Returns the number of items in the index.
#'
#' \code{$setVerbose()}
#' If \code{1} then messages will be printed during processing.
#' If \code{0} then messages will be suppressed during processing.
#'
#' \code{$setSeed()}
#' Set random seed for annoy (integer).
#'
#' @examples
#' library(RcppAnnoy)
#'
#' # BUILDING ANNOY INDEX ---------------------------------------------------------
#' vector_size <- 10
#' a <- new(AnnoyEuclidean, vector_size)
#'
#' a$setSeed(42)
#'
#' # Turn on verbose status messages (0 to turn off)
#' a$setVerbose(1)
#'
#' # Load 100 random vectors into index
#' for (i in 1:100) a$addItem(i - 1, runif(vector_size)) # Annoy uses zero indexing
#'
#' # Display number of items in index
#' a$getNItems()
#'
#' # Retrieve item at postition 0 in index
#' a$getItemsVector(0)
#'
#' # Calculate distance between items at postitions 0 & 1 in index
#' a$getDistance(0, 1)
#'
#' # Build forest with 50 trees
#' a$build(50)
#'
#'
#' # PERFORMING ANNOY SEARCH ------------------------------------------------------
#'
#' # Retrieve 5 nearest neighbors to item 0
#' # Returned as integer vector of indices
#' a$getNNsByItem(0, 5)
#'
#' # Retrieve 5 nearest neighbors to item 0
#' # search_k = -1 will invoke default search_k value of n_trees * n
#' # Return results as list with an element for distance
#' a$getNNsByItemList(0, 5, -1, TRUE)
#'
#' # Retrieve 5 nearest neighbors to item 0
#' # search_k = -1 will invoke default search_k value of n_trees * n
#' # Return results as list without an element for distance
#' a$getNNsByItemList(0, 5, -1, FALSE)
#'
#'
#' v <- runif(vector_size)
#' # Retrieve 5 nearest neighbors to vector v
#' # Returned as integer vector of indices
#' a$getNNsByVector(v, 5)
#'
#' # Retrieve 5 nearest neighbors to vector v
#' # search_k = -1 will invoke default search_k value of n_trees * n
#' # Return results as list with an element for distance
#' a$getNNsByVectorList(v, 5, -1, TRUE)
#'
#' # Retrieve 5 nearest neighbors to vector v
#' # search_k = -1 will invoke default search_k value of n_trees * n
#' # Return results as list with an element for distance
#' a$getNNsByVectorList(v, 5, -1, TRUE)
#'
#'
#' # SAVING/LOADING ANNOY INDEX ---------------------------------------------------
#'
#' # Create a tempfile, replace with a local file to keep
#' treefile <- tempfile(pattern="annoy", fileext="tree")
#'
#' # Save annoy tree to disk
#' a$save(treefile)
#'
#' # Load annoy tree from disk
#' a$load(treefile)
#'
#' # Unload index from memory
#' a$unload()
NULL

## ensure module gets loaded
loadModule("AnnoyAngular", TRUE)
loadModule("AnnoyEuclidean", TRUE)
loadModule("AnnoyManhattan", TRUE)
loadModule("AnnoyHamming", TRUE)
